### PR TITLE
[verifier][adapter] Fix issues with immutable txn context

### PR DESCRIPTION
**Pull request by tnowacki**
- Fix issue where the adapter would look for mutable results for an immutable txn context
- Added back the restriction that init must have a mutable txn context for simplicity

### CODE REVIEW

File: crates/sui-adapter/src/adapter.rs [hunk 1]
```diff
@@ -425,15 +425,16 @@ fn init_modules<
             args.push(bcs_char_type_value);
         }
         args.push(ctx.to_vec());
-        let has_ctx_arg = true;
+        // init must have a mutable
+        let ctx_is_mut = Some(true);
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:429]: If we want to allow immutable txn context for `init`, we would have to change this to check the signature of the function for the information. 

That being said, I'm not sure what good an `init` function does without being able to create new objects

**Review by emmazzz — APPROVED** [review]
LGTM

**Comment by tnowacki** [comment]
I'll wait for @sblackshear's input on the `init` change before landing

File: crates/sui-adapter/src/adapter.rs (hunk 1 above)
Comment (sblackshear) [reply crates/sui-adapter/src/adapter.rs:429]: I think you're right that just about any useful init will have `&mut TxContext` (though I could imagine e.g., emitting an event that contains the epoch where a module was published, which wouldn't need `&mut`).

However, I'd prefer to have the same rules everywhere just so we don't have to explain to programmers that `init` is a special case, even if it makes the code here a bit more complicated. I think it's simpler to explain that you can use both `&TxContext` and `&mut TxContext` everywhere.

File: crates/sui-verifier/src/entry_points_verifier.rs [hunk 2]
```diff
@@ -252,18 +254,26 @@ fn is_primitive(
     }
 }
 
-pub fn is_tx_context(view: &BinaryIndexedView, p: &SignatureToken) -> bool {
+// Returns Some(is_mut) if the type is a reference to the TxnContext. is_mut being true with
+// a MutableReference, and false otherwise.
+// Returns None for all other types
+pub fn is_tx_context(view: &BinaryIndexedView, p: &SignatureToken) -> Option<bool> {
```
Comment (sblackshear) [inline crates/sui-verifier/src/entry_points_verifier.rs:260]: I think an `enum` or newtype would be clearer here --I keep incorrectly thinking the `tx_context_mut.is_some()`  check elsewhere is true only when you get a mutable txcontext.

**Comment by sblackshear** [comment]
Thanks for fixing this, I should have added adapter tests...

File: crates/sui-verifier/src/entry_points_verifier.rs (hunk 2 above)
Comment (tnowacki) [reply crates/sui-verifier/src/entry_points_verifier.rs]: It's maybe a bit obtuse not using an `Option`, but at least we have to manually check it each time, and I prefer that over hiding cases behind methods

File: crates/sui-adapter/src/adapter.rs (hunk 1 above)
Comment (tnowacki) [reply crates/sui-adapter/src/adapter.rs]: Fair enough, reverted

**Pull request merged**